### DISCUSSION TITLE
Interest in an official docker?

### DISCUSSION THREAD

**Discussion by Bockiii** [jamulussoftware/jamulus#2133] [Ideas]
Hi,

I can see that there is an inofficial docker with ~100k pulls on dockerhub.

Are you interested in an official docker container provided by you? I already started a PR on this but since I wanted to integrate it to into your release workflow, it's taking a bit longer (it could be done quicker, but then it's not integrated). But before I spend more time on it, I wanted to know if you are even interested in that.

So it would require you create an account on dockerhub and in addition to that, it would also create the image on the github repo.

I've set this up for another application with a similar release workflow than yours and would configure it to be "latest" (actually last commit), stable (last released version) and versionized tags (e.g. jamulus:3.8.1).

**Comment by rdica** [comment]
Isn't dockerhub now limiting pulls and integrations for non-commercial use?

**Reply by Bockiii** [reply]
Correct, thats why you can offer both with a single build process. Offering it on dockerhub doesn't bring you any downsides, but some heavier users might run into the limits.

Also, autobuilding isnt supported anymore, but thats done via github action anyways.

**Comment by Bockiii** [comment]
The other thing I've setup: https://github.com/RSS-Bridge/rss-bridge/blob/master/.github/workflows/dockerbuild.yml

**Reply by Bockiii** [reply]
This builds for x64, armv7 and armv8 btw which would also be directly available to you. so people can run their server on a pi with very little effort.

**Comment by gilgongo** [comment]
We're not a very big team on Jamlus, so I'd be a bit wary of having to maintain a Docker build alongside the other ones, particularly if it involves a 3rd party account to keep hold of. Not sure what others think.

**Reply by pljones** [reply]
I wouldn't object to someone contributing a Dockerfile.  I think, once working, it should be fairly stable.

I would probably not be comforable about having the build process push a Docker image to the hub.

**Reply by Bockiii** [reply]
It doesnt necessarily need a 3rd party account, as you could skip dockerhub all together. It's just "the" repo to search for things, so if someone would look for a docker image of jamulus, they would search for it on dockerhub and not go to the repo to see if there is an image in the container repo on github. So you could keep the image on github if you want to, it just means less visibility.

As for the build process pushing: That's very common practice today. Since dockerhub limited the use of their build system (because of bad actors who did crypto mining using the build system), most docker containers moved their build systems to github/gitlab and then push to dockerhub.

I also agree: If you dont switch up something dramatic in the background, the docker image should be pretty stable.

**Comment by gilgongo** [comment]
BTW There is [this ticket](https://github.com/jamulussoftware/jamuluswebsite/issues/489) if anyone wants to maintain a page summarising the many and various methods of Linux installation for Jamulus. Over 15 years it's built up quite a few (although I don't think Linux use in general has been more than 5% of downloads).

**Reply by Bockiii** [reply]
While I agree (and already assumed) that the userbase for the Client side of it on Linux is probably small (linux is notoriously a pita with audio interfaces and good audio processing in general), I would also assume that most servers are probably run on linux machines.

this would enable users to spin it up on their QNAP/synology nas in 3 minutes.

If I were to do this, I would probably base most of the actual dockerfile off of the one provided by @grundic in https://github.com/grundic/jamulus-docker and then add the auto build with version logic from my previous setups. I would also include it in the already present build script since you are already providing things like clean version number and so on in there.

**Reply by Bockiii** [reply]
maybe grundic wants to join in and we do it together :)

**Reply by ghost** [reply]
My experience with Linux and Windows is that Windows programs run faster and stabler with the Wine Windows emulator under Linux than they do on native Windows. Experience Linux, Experience Freedom.

**Reply by gilgongo** [reply]
Yes, perhaps about 75% of public servers are currently Linux.

From a Linux server admin's POV though, the advantage of installing a Docker image vs running a .deb seems fairly marginal perhaps (particularly as there aren't many dependencies for headless servers, and apt brings them in automatically). 

So the question is whether the advantages of an alternative installation method like Docker are enough for us to have the build process push that to GH. I think the overhead of maintenance over time makes it feel like we can't really take that on at the moment. There are also things like Flatpack and snaps that people might prefer, but I doubt we'd want to make builds for them as well.

No problem linking to any 3rd party maintained projects from the aforementioned Knowledge Base article though.

**Reply by Bockiii** [reply]
Portability, preconfiguration, updates, there are so many advantages of containers over installing a deb that it would make this thread too long :)

But I can hear from the tone that you are not overly interested in providing an official image and that's totally fine :)

So you can close the conversation and people can just use the docker image from grundic.

**Reply by gilgongo** [reply]
OK so maybe this is a better way of framing the question then, given the below points:

1. We a wary of having to maintain too many autobuilds (GH changes with permissions, updates, library versions, etc. etc.) so would prefer just one official installation method per platform.

2. We currently say that a .deb binary install is the only official method of installing Jamulus on Linux. We also plan to host a PPE for this in due course too.

3. Perhaps 75% of public servers run Linux, few clients (appro. 5% we think) are on Linux.

So the question might not therefore be _"Are we interested in adding an official docker image?"_ But _"Do we think a .deb is the right call?"_ 

Should we instead move to Docker images for the official Linux install method? The answer might be "yes" if we think the advantages of a Docker image are objectively better **for users of Jamulus** (note my emphasis)  than a .deb.

**Reply by grundic** [reply]
I've sent an application to the Docker Open Source Program, which should give us access to all necessary automation for building package. They're a bit busy and slow on reply, but I hope they'll approve it. 

@Bockiii, if you have any ideas about improving current Docker setup or add some feature -- you are more then welcome to send PR 😉

**Reply by Bockiii** [reply]
Grundic: I was actually in the process of creating a pr on your image when I thought that it may make sense to do an "official" one directly. 

The problem of being disconnected from the source is always build planning. How are you aware that a new version was released? So you start building up schedules that compare release versions and so on. It is way easier to trigger that in the source during the actual release.

So I'll open a pr later with a first idea.

**Reply by pljones** [reply]
> I've sent an application to the Docker Open Source Program, which should give us access to all necessary automation for building package. They're a bit busy and slow on reply, but I hope they'll approve it.

This doesn't inspire confidence in the approach if it's going to be an on-going dependency.  Can you explain the comment further?

**Reply by Bockiii** [reply]
That's the docker program to get builds working again on their infrastructure. We wouldn't need that as building would be done on github.

Docker the company has done some questionable business moves in the last years but that doesn't change the value of images and containers :)

People can also use other means to run/orchestrate containers like podman or kubernetes.

**Reply by Bockiii** [reply]
Is there a way to provice a single configuration file to set all the things that I could set through cli options? 

I can see that grundic used entrypoint extensions, but I would rather not use those. Either env variables or a config file :)

**Reply by ann0see** [reply]
> But I can hear from the tone that you are not overly interested in providing an official image and that's totally fine :)

The main reason is probably maintenance cost - and https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#keep-it-simple-and-stupid-and-do-one-thing-and-do-it-well

> The problem of being disconnected from the source is always build planning. How are you aware that a new version was released?

Polling the GH API, using IFTTT actions, cron (https://jasonet.co/posts/scheduled-actions/),... are all automated methods. OFC. they need some setup.

I mean, it should be possible to "link" and automate builds. 

Something like:

```
Run every 2 days on GitHub via schedueled-actions:
  if (dockerRepo.getCurrentRelease() != GitHub.API.getLatestReleaseOf(jamulussoftware/jamulus)) {
    dockerRepo.getNewSource()
    dockerRepo.build()
    dockerRepo.publishToDockerHub()
  }

dockerRepo.getCurrentRelease():
  // probably implemented with some bash script which gets the latest release from DockerHub
```

**Reply by Bockiii** [reply]
You missed the second part of the quote: 
> So you start building up schedules that compare release versions and so on. It is way easier to trigger that in the source during the actual release.

Using IFTTT, cron and the likes is exactly that. Yes it works (grundic is doing it like that and many other docker images do that), but it's detached and always has the "inofficial" character, which could be a reason for system operators not to use it.

> The main reason is probably maintenance cost

I think that depends on what you count against each other. Cost of installing (and sometimes building), updating, configuring and maintaining of those 75% of jamulus servers is currently offloaded to the server providers. If you can lessen the effort for those, the ROI is quite big, compared to the maintenance effort of the docker image (since you usually build images to be code stable through releases). 

Also, if you run it on a schedule, you also "cost" github processing time without a reason (since 99% of runs will return with "no release" and gh still had to start a vm and do things).

**Reply by ann0see** [reply]
Hmm. Fair points.

Maybe we can agree on the following:
1. we move grundic‘s repo to the jamulussoftware repo
2. We give you both maintenance access to the repo
3. You‘re still responsible for the code and everything around docker; however we formally support your docker image. Actually, the signed macOS build is also not 100% team-owned since emlyn is in charge of it.

**Reply by ann0see** [reply]
But that’s up to @jamulussoftware/maindevelopers

**Reply by Bockiii** [reply]
We would need to do changes to the current dockerfile because of the different build process but in general, sure that would work. It wouldn't help with the detached situation though :)

**Reply by Bockiii** [reply]
I have used most of grundics dockerfile as the base for the "new" image.

https://github.com/Bockiii/jamulus/pkgs/container/jamulus

Right now it's basically the same functionality (the only thing I really changed was upping the alpine version and using the lsio alpine for the runner since it provides things like s6 etc) except that it is already available in multi-arch.

The next thing I want to do is provide a different option to do the settings than entrypoint extensions, but that will require some scripting and some thoughts as I understand that there is no config-file I could just parse.

**Reply by melcon** [reply]
> Right now it's basically the same functionality (the only thing I really changed was upping the alpine version and using the lsio alpine for the runner since it provides things like s6 etc) except that it is already available in multi-arch.

Do you know by any chance if that lsio alpine has support for aws-efs-utils? Or have you tried to add support for it yourself?

**Reply by Bockiii** [reply]
Is there a reason why you're not mounting the efs drive on the host and then just mount the volume? I assume you want to share it between containers, so... why not mount on the host?

A container is usually a do-one-thing kind of deal. So if you need extra things to mount something, it would make more sense to use a container thats specific to do that mounting and share a volume between that container and the application container.

To give you an example: You dont add a VPN capability to a torrent-client container. You create a vpn container and route the torrent-client container through that vpn container.

**Comment by gene96817** [comment]
I don't have any hands-on docker experience. I would be interested in running 4+ server instances in a single (hw) server. It would be awesome to know how to size each docker instance for a max number of participants (where the max is not all the same). Consider four containers, each for 8 participants, and one container for 40 participants.

**Reply by Bockiii** [reply]
running 4 instances would be no problem. You can assign different ports to every container.

As for sizing: You dont need to size containers. They take as much as they need. It's not like running a VM. Thats one of the benefits of containerization.

**Reply by gilgongo** [reply]
@gene96817 

> I would be interested in running 4+ server instances in a single (hw) server.

Just to be clear - unless you tell your users about the port number to connect their clients on, you will have to have a machine with multiple network interfaces to do this.

**Reply by Bockiii** [reply]
install docker, install docker-compose, save the following code to a file called "docker-compose.yml", type "docker-compose up" in that folder and viola, 4 running instances with a different port (1001-1004). No further installation needed, no QT juggling, nothing.

And when you want to kill all the servers, "docker-compose down" and everything is "cleaned up".

```
version: "3.7"
services:  
  jamulus1:
    container_name: jamulus1
    image: grundic/jamulus
    restart: always
    ports:
      - "1001:22124/udp"
    environment:
      TZ: "America/Los_Angeles"  
    entrypoint:
      - "Jamulus"
      - "--server"
      - "--nogui"
      - "--welcomemessage"
      - "Welcome to Jamulus 1"
      - "--numchannels"
      - "16"
  jamulus2:
    container_name: jamulus2
    image: grundic/jamulus
    restart: always
    ports:
      - "1002:22124/udp"
    environment:
      TZ: "America/Los_Angeles"  
    entrypoint:
      - "Jamulus"
      - "--server"
      - "--nogui"
      - "--welcomemessage"
      - "Welcome to Jamulus 2"
      - "--numchannels"
      - "16"
  jamulus3:
    container_name: jamulus3
    image: grundic/jamulus
    restart: always
    ports:
      - "1003:22124/udp"
    environment:
      TZ: "America/Los_Angeles"  
    entrypoint:
      - "Jamulus"
      - "--server"
      - "--nogui"
      - "--welcomemessage"
      - "Welcome to Jamulus 3"
      - "--numchannels"
      - "16"
  jamulus4:
    container_name: jamulus4
    image: grundic/jamulus
    restart: always
    ports:
      - "1004:22124/udp"
    environment:
      TZ: "America/Los_Angeles"  
    entrypoint:
      - "Jamulus"
      - "--server"
      - "--nogui"
      - "--welcomemessage"
      - "Welcome to Jamulus 4"
      - "--numchannels"
      - "16"
```

**Reply by gene96817** [reply]
> running 4 instances would be no problem. You can assign different ports to every container.
> 
> As for sizing: You dont need to size containers. They take as much as they need. It's not like running a VM. Thats one of the benefits of containerization.

How do you prevent running our of resources? Consider a situation where there are one 25-person container and six 8-person container. If the 25-person container is in use, is there a way to prevent three of 8-person container to be used? It is possible this is not the right way to ask the question. I hope you can get the meaning of the question.

**Reply by Bockiii** [reply]
No, you can't prevent that. It's also not the correct question tbh :D Containerization does not replace resource planning. You could go and use an orchestration tool like kubernetes to spread resources over multiple nodes, but I don't think that's what you meant. 

If your machine is only capable of hosting 20 people, the amount of instances doesnt really matter. You can have one container with 20 people it it or 20 containers with 1 person in it. The benefit of containers is that you dont need to install anything on your host apart from the docker runtime. No dependency nightmares, no apps/libs that fill up your host system and so on. It doesn't magically improve your host machines performance cap :)

thats the answer, here is a little deeper dig into the orchestration part, you can skip this if you dont care about kubernetes:
Imagine you would be a person that wants to host jamulus servers "for everyone", meaning: Every person that wants to join a server shall be able to (and also lets assume money doesnt matter). You could set up a kubernetes network with a single app, Jamulus. You get one "node" (thats a server that takes commands from kubernetes to run an app) in an amazon datacenter and through performance testing etc you find out that it's capable of hosting 50 people. You tell kubernetes to generate 10 people instances and always have 5 servers empty (for new sessions). You also tell it to only use a node for a maximum of 5 instances per node since in theory, 5 instances with 10 people each could be on one node. You spin up the system, kubernetes starts 5 instances of the docker container, configured for 10 people.

Person1 connects to one of those servers. Kubernetes detects that your configured value of empty servers is not met anymore (as now we have 4 empty servers) and spins up another instance (so now you have 6 instances running). Since you have reached the maximum of instances on one node, kubernetes sets up another node in the amazon datacenter (again... money doesnt matter ;) ) and spins up the container on node2. People connect to instance 2 and 3, so kubernetes spins up 2 more instances on node2 (now you have 8 instances running, 3 with people on it, 5 empty). Now all the people from instance 1 leave, bringing your "empty container count" to 6, which is not the number configured, so kubernetes shuts down an empty instance (not neccessarily instance1, it does some node management as well so you dont have 5 nodes with 1 instance each).

All of that is configurable in a managed amazon kubernetes setup in less than a day and can theoretically run forever. You can do auto-patching (so when there is a new version of the image, either cycle through the instances, so when a new instance is build, use the new version or just kill all open instances and restart or just fix the version that you want and not upgrade at all etc) and so on and so on.

This would be a very basic setup, you can do all kinds of fancy things with kubernetes in the sense of load management and so on.

**Reply by gene96817** [reply]
Thanks for the long response. Too bad a container does not enforce a resource constraint. I don't expect containers to create more capacity. I was hoping that a number of containers can be created so that groups with low utilization can have their own container without worrying about crashing the server when all the resources are used. Of course, scheduling and coordination is a different function. 

Using Kubernetes to spin up new instances doesn't preserve the behavior of a public server. I would like to see public servers available and listed on the directory server. Since I expect the public server has a low utilization rate, the (hw) server can keep a number of instances running with low resource utilization because utilization is low. (I know this creates new challenges. I consider them good side effects.)

**Reply by Bockiii** [reply]
> Too bad a container does not enforce a resource constraint.

Thats not what I said. You can set container max-usages for CPU, RAM, disk space and so on, but that's not really what you would need to solve your initial problem.

From your theroretical setup, I assume you know you have a limit of people that your hardware (cpu+ram) supports. Lets say that's 25. what you could do is configure 5 containers (configured to host maximum of 5 people) to use a maximum of 0.2 of your CPU (so 20%) and ram. That way, no container on it's own can use more than 20% of the hardware and all containers can run at full capacity (5*5, 25 people). The problem with this setup is: The max-cpu setting will not really do anything, as you have already limited the people to 5 per instance. If you set the max-people value higher (lets say 10 people per instance), you will run into problems with your 20% setting per instance (as your hardware is only capable of hosting 25 all together ... see where I'm going?)

As for crashing the host because of docker container utilization: I've honestly never heard of that, but of course it could happen. I've been running containerization on metal for a few years now (from single docker runs to docker-compose to k3s) and never had a host crash when a container crashed (I'm running plex so yes... containers have crashed :D ).

As for the public hosting: that should 100% be doable with a reverse proxy and a service like cloudflare dns. I have a domain, dns managed by cloudflare, traefik reverseproxy configured and a cloudflare-helper-container that immediately creates subdomains for containers I spin up. If I start a jamulus instance, its instantly available under jamulus.mydomain.com . You can do jamulus1.mydomain, jamulus2.mydomain etc.

**Reply by gene96817** [reply]
@Bockiii    Thank you for this dialog. 

You understand my scenario. I would add one twist. There are more containers and it is assumed that not all containers can have users, thus not exceeding the hardware resource limits.

Clearly, I don't have any experience with docker.  This topic is my next Jamulus frontier. Is anyone interested in working with me on testing this concept?

**Reply by Bockiii** [reply]
If you want a good guide to start, follow this:

https://www.smarthomebeginner.com/traefik-2-docker-tutorial/

If you read it carefully (and follow his links to topics that dont tell you anything) you can get a lot of this running relatively quickly (minus the kubernetes orchestration part. but thats for V2 I would say ;) ).

**Reply by gilgongo** [reply]
@gene96817 In your case it perhaps might be better to run the servers just using systemd with nice and ionice settings. Resource allocations would be calculated from a combination of the different processes' niceness values, plus other things like the I/O of each, not just a simple percentage. Like @Bockiii says, the main advantage of using docker is so that you don't have to worry about installing libraries and things on systems that don't have them - it's not about process tuning.

**Comment by gene96817** [comment]
@gilgongo    What do I know about each instance? What files can be shared between instance and which files have to be unique?  Thanks.

**Reply by gilgongo** [reply]
I've not tried it (and this doesn't involved Docker) but if you create a bash script for each of your servers containing the desired startup and nice value options, eg: `jamulusA.sh`, `jamulusB.sh` etc. each containing something like:

```
 nice -n [niceValue] /usr/local/bin/jamulus-headless  \
        --directoryserver genreServer:port \
        --serverinfo "yourServerName;yourCity;[country ID]"
	... [etc]
```

then have a systemd "template" unit file with a `%i` variable in it:

`ExecStart=/bin/sh -c 'exec /usr/local/bin/jamulus%i.sh'`

You can then control them with `systemctl [start/stop] jamulus@A.service`  `systemctl [start/stop] jamulus@B.service` etc. The bit directly after the `@` is used by the `%i` variable in the unit file. A more [detailed illustration is here](https://www.stevenrombauts.be/2019/01/run-multiple-instances-of-the-same-systemd-unit/).

**Reply by pljones** [reply]
I do something much more complex in https://github.com/pljones/jamulus-systemd
